flink-datastream/src/main/java/org/apache/flink/datastream/impl/extension/eventtime/functions/ExtractEventTimeProcessFunction.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.datastream.impl.extension.eventtime.functions;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.operators.ProcessingTimeService;
+import org.apache.flink.api.common.watermark.WatermarkDeclaration;
+import org.apache.flink.api.common.watermark.WatermarkManager;
+import org.apache.flink.configuration.PipelineOptions;
+import org.apache.flink.datastream.api.common.Collector;
+import org.apache.flink.datastream.api.context.PartitionedContext;
+import org.apache.flink.datastream.api.extension.eventtime.EventTimeExtension;
+import org.apache.flink.datastream.api.extension.eventtime.strategy.EventTimeWatermarkStrategy;
+import org.apache.flink.datastream.api.function.OneInputStreamProcessFunction;
+import org.apache.flink.util.clock.Clock;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** A specialized process function designed for extracting event timestamps. */
+public class ExtractEventTimeProcessFunction<IN>
+        implements OneInputStreamProcessFunction<IN, IN>,
+                ProcessingTimeService.ProcessingTimeCallback {
+
+    /** User-defined watermark strategy. */
+    private final EventTimeWatermarkStrategy<IN> watermarkStrategy;
+
+    /** The maximum timestamp encountered so far. */
+    private long currentMaxEventTime = Long.MIN_VALUE;
+
+    /** The maximum timestamp encountered so far. */
+    private long lastEmittedEventTime = Long.MIN_VALUE;
+
+    /**
+     * The periodic processing timer interval; if not configured by user in {@link
+     * EventTimeWatermarkStrategy#getPeriodicWatermarkInterval()}, it will default to the value
+     * specified by {@link PipelineOptions#AUTO_WATERMARK_INTERVAL}.
+     */
+    private long periodicTimerInterval = 0;
```

Comment:
```suggestion
    /**
     * The periodic processing timer interval; if not configured by user in {@link
     * EventTimeWatermarkStrategy}, it will default to the value
     * specified by {@link PipelineOptions#AUTO_WATERMARK_INTERVAL}.
     */
    private long periodicTimerInterval = 0;
```